lib/msf/core/module.rb

```diff
@@ -151,7 +151,7 @@ def replicant
       obj = self.clone
       self.instance_variables.each { |k|
         v = instance_variable_get(k)
-        v = v.dup rescue v
+        v = (v.is_a?(Rex::SharedResource) ? v : v.dup) rescue v
```

Comment:
Unlike the `Ref` approach, this doesnt leave any indication as to how many duplicates we have unless the results of the `:dup` are placed into a data structure which will hold reference to them.

Reply:
Thinking about this some more, it occurs that maybe we do want a data structure there instead - like the monitor threads and low-level plumbing in Rex, it might make sense to have the replicant-creating class hold reference and be responsible for cleanup

Reply:
Ah, right, the `deref` thing - i was thinking about that and it dawns on me that we can have the MetasploitModule deref anything ref'd in it when it closes as a finalizer (hopefully not an actual finalizer as GC interactions are gross and VM specific - years ago i had to carve continuations out of Msf because i was porting it to Rubinius). This would bring the refcount to 1 when the original module was done scanning and let it dispose of the object in _its_ finalizing call. That would allow us to refcount up and safely refcount down to destructor.
Also to clarify a bit, the way Ref works is weird - @jmartin-r7 and i ran into this too: ref doesn't return a reference, it increments a ref counter on the object containing the mixin and returns the object itself just like you guys are doing here. I think Ref might also be extensible in a few ways, one of which being to have a GC of sorts for Ref-mixed objects that get "forgotten" by framework as was happening to the `self.service` call site in the original module when the duplicates were failing to bind the same socket